# Tugas Besar Analisa Data (Kelompok 5) DS-47-01



**Model Regresi Linear Bertahap untuk Memprediksi Gaji Teknisi**

Muhammad Zikra Al Rizkya Adler - 103052300076

Taufik Qurohman - 103052300055

Hauzan Rafi Attallah - 103052330011

**1. Pengolahan Data**

In [ ]:
import pandas as pd
import numpy as np
import scipy.stats as st
import math

data = pd.read_excel('Kel-5-Data-Gaji-teknisi.xlsx')

In [ ]:
data

,GraduationYear,English,Logical,ComputerProgramming,Salary
0,2013,650,665,485,445000
1,2014,440,435,365,110000
2,2013,675,620,655,420000
3,2012,575,495,315,200000
4,2013,535,595,455,440000
5,2013,370,470,465,105000
6,2014,510,555,525,195000
7,2014,500,410,385,200000
8,2013,675,630,405,335000
9,2013,395,565,495,300000


In [ ]:
kolom = [data['GraduationYear'], data['English'], data['Logical'], data['ComputerProgramming'], data['Salary']]
n = len(kolom[-1])

n

50

Variabel Bebas (Independent Variables):

*   X₁ = GraduationYear
*   X₂ = English
*   X₃ = Logical
*   X₄ = ComputerProgramming

Variabel Terikat (Dependent Variable):
*  Y = Salary




**2.  Menghitung korelasi**

**I. Membuat rumus menghitung korelasi**

In [ ]:
def calculate_correlation(x, y):
   # Rumus korelasi: r(Xi,Xj) = [Σ(Xi×Xj) - (ΣXi×ΣXj)/n] / √[(ΣXi² - (ΣXi)²/n) × (ΣXj² - (ΣXj)²/n)]
   n = len(x)

   # Komponen dasar
   sum_xi = sum(x)
   sum_xj = sum(y)
   sum_xi_squared = sum([xi**2 for xi in x])
   sum_xj_squared = sum([xj**2 for xj in y])
   sum_xi_xj = sum([x[i] * y[i] for i in range(n)])

   # Pembilang: Σ(Xi×Xj) - (ΣXi×ΣXj)/n
   numerator = sum_xi_xj - (sum_xi * sum_xj) / n

   # Penyebut: √[(ΣXi² - (ΣXi)²/n) × (ΣXj² - (ΣXj)²/n)]
   denominator_xi = sum_xi_squared - (sum_xi**2) / n
   denominator_xj = sum_xj_squared - (sum_xj**2) / n
   denominator = math.sqrt(denominator_xi * denominator_xj)

   # Return korelasi atau 0 jika pembagi nol
   if denominator == 0:
       return 0
   else:
       return numerator / denominator

**II Perhitungan Detail**

In [ ]:
variables = ['GraduationYear', 'English', 'Logical', 'ComputerProgramming', 'Salary']
n = len(data)

print(f"n = {n}")
for i, var in enumerate(variables):
    col_data = kolom[i].values
    print(f"{var}: Σx = {sum(col_data):.2f}, Σx² = {sum([x**2 for x in col_data]):.2f}")

n = 50
GraduationYear: Σx = 100640.00, Σx² = 202568250.00
English: Σx = 23523.00, Σx² = 11784007.00
Logical: Σx = 24680.00, Σx² = 12548200.00
ComputerProgramming: Σx = 23330.00, Σx² = 11334050.00
Salary: Σx = 14500000.00, Σx² = 5107950000000.00


Data berikut merupakan hasil perhitungan awal (statistik deskriptif) yang akan menjadi dasar untuk analisis lebih lanjut dalam penelitian atau studi regresi. Perhitungan awal ini mencakup jumlah total (Σx) dan jumlah kuadrat (Σx²) dari masing-masing variabel yang diamati terhadap 50 responden (n = 50).

**III. Tabel Korelasi**

In [ ]:
correlation_table = []

for i in range(len(kolom)):
    row = []
    for j in range(len(kolom)):
        if i == j:
            corr = 1.000
        else:
            x_data = kolom[i].values
            y_data = kolom[j].values
            corr = calculate_correlation(x_data, y_data)
        row.append(corr)
    correlation_table.append(row)

# Tampilkan tabel
header = "Variable".ljust(12)
for var in variables:
    header += var[:8].rjust(10)
print(header)

for i, var in enumerate(variables):
    row_str = var[:11].ljust(12)
    for j in range(len(variables)):
        row_str += f"{correlation_table[i][j]:.3f}".rjust(10)
    print(row_str)

Variable      Graduati   English   Logical  Computer    Salary
GraduationY      1.000    -0.081     0.230    -0.097    -0.244
English         -0.081     1.000     0.514     0.407     0.194
Logical          0.230     0.514     1.000     0.453     0.081
ComputerPro     -0.097     0.407     0.453     1.000     0.111
Salary          -0.244     0.194     0.081     0.111     1.000


1. Korelasi tertinggi dalam dataset terjadi antara variabel Logical dan English, dengan nilai korelasi 0.514, menunjukkan hubungan positif yang cukup kuat.

2. Korelasi antara seluruh variabel prediktor terhadap variabel Salary sangat rendah. Nilai korelasi tertinggi hanya 0.194 (antara English dan Salary), menunjukkan bahwa tidak ada hubungan linier yang kuat antara keempat variabel prediktor dan gaji.

3. Korelasi antara Graduation Year dan Salary sebesar -0.244, yang berarti semakin baru tahun kelulusannya, cenderung semakin rendah gajinya.

**IV. Interpretasi Hasil**

In [ ]:
print("\nKorelasi signifikan (|r| ≥ 0.3):")
for i, var1 in enumerate(variables):
    for j, var2 in enumerate(variables):
        if i < j:
            corr = correlation_table[i][j]
            if abs(corr) >= 0.3:
                print(f"{var1} vs {var2}: r = {corr:.3f}")


Korelasi signifikan (|r| ≥ 0.3):
English vs Logical: r = 0.514
English vs ComputerProgramming: r = 0.407
Logical vs ComputerProgramming: r = 0.453


Korelasi yang signifikan terjadi di antara kemampuan kognitif (Bahasa Inggris, Logika, dan Pemrograman).

Hal ini menunjukkan bahwa keterampilan akademik ini saling berkaitan dan dapat saling memperkuat.

Namun, tidak ada korelasi kuat yang signifikan dengan variabel Salary, sehingga kemampuan ini belum tentu berdampak langsung terhadap gaji awal, setidaknya dalam data ini.

**3. Stepwise**

**I. Membuat rumus**

In [ ]:
# Buat rumus-rumus

# JKR sederhana
def Jumlah_Kuadrat_Regresi(col1, col2, n):
  x = np.sum(col1)
  y = np.sum(col2)
  x_bar = x / n
  y_bar = y / n
  xy = np.sum(col1 * col2)
  x2 = np.sum(col1 ** 2)

  b = (n * xy - x * y)/(n * x2 - x**2)
  a = y_bar - b * x_bar
  yi = []
  for i in range(n):
    yi.append(a + b * col1[i])
  hasil = np.sum((yi - y_bar)**2)
  return hasil

# JKR berganda
def Jumlah_Kuadrat_Regresi_Berganda(xs, y):
  xs = [np.array(x) for x in xs]
  X_mat = np.column_stack([np.ones(len(y))] + xs)
  beta = np.linalg.inv(X_mat.T @ X_mat) @ (X_mat.T @ y)
  Y_pred = X_mat @ beta
  JKR = np.sum((Y_pred - np.mean(y))**2)
  return JKR, beta

# JKT
def Jumlah_Kuadrat_Total(col1, n):
  y = np.sum(col1)
  y_bar = y / n
  hasil = 0
  for i in range(n):
    hasil =  (col1[i] - y_bar)**2 + hasil

  return hasil

JKT = Jumlah_Kuadrat_Total(kolom[-1], n)
print("nilai JKT =",f"{JKT:.4f}")

nilai JKT = 902950000000.0000


Nilai JKT yang besar (902.95 miliar) menunjukkan bahwa nilai gaji (Salary) memiliki keragaman yang tinggi (varian besar) dalam dataset ini

**II. Uji satu satu**

In [ ]:
# Uji satu satu
# Menghitung nilai R tiap peubah
JKR_sederhana = []
for i in kolom[:-1]:
  hasil = Jumlah_Kuadrat_Regresi(i, kolom[-1],n)
  JKR_sederhana.append(hasil)

# Melihat masing masing nilai R
print("")
print("TAHAP 1 - Uji peubah tunggal")
for i, JKR in enumerate(JKR_sederhana):
  print(f"R X{i+1} =" ,f"{JKR:.4f}")

# Mencari peubah yang memiliki nilai R terbesar
print("")
print("TAHAP 2 - cari R terbesar")
idx_terbesar = np.argmax(JKR_sederhana)
print(f"R terbesar adalah X{idx_terbesar+1} = {JKR_sederhana[idx_terbesar]:.4f}")

# Membuat array untuk menyimpan nilai nilai
variable_terpilih = []
variable_sisa = list(set(range(4)) - set(variable_terpilih))
JKR_sekarang = JKR_sederhana[idx_terbesar]

# Menghitung nilai F hitung dan F tabel
print("")
print(f"TAHAP 3 - hitung nilai F dari X{idx_terbesar+1}")
df = n - (len(variable_terpilih)+ 1 + 1)
S2 = (JKT - JKR_sekarang) / df
F_hitung = JKR_sekarang / S2
F_tabel = st.f.ppf(0.95,1,df)
print("R =",f"{JKR_sederhana[idx_terbesar]:.4f}")
print("S kuadrat =",f"{S2:.4f}")
print("nilai F hitung =",f"{F_hitung:.4f}")
print("nilai F tabel =", f"{F_tabel:.4f}")

# Keputusan apakah peubah masuk ke dalam model atau tidak
if F_hitung > F_tabel:
  print(f"peubah X{idx_terbesar+1} masuk dalam model")
  variable_terpilih.append(idx_terbesar)
else:
  print(f"peubah X{idx_terbesar+1} tidak masuk dalam model")


TAHAP 1 - Uji peubah tunggal
R X1 = 53710775862.0677
R X2 = 34089083106.4673
R X3 = 5854645065.4373
R X4 = 11103308928.5077

TAHAP 2 - cari R terbesar
R terbesar adalah X1 = 53710775862.0677

TAHAP 3 - hitung nilai F dari X1
R = 53710775862.0677
S kuadrat = 17692483836.2069
nilai F hitung = 3.0358
nilai F tabel = 4.0427
peubah X1 tidak masuk dalam model


Dari uji peubah tunggal, tidak ada peubah yang secara signifikan memengaruhi Salary jika dilihat satu per satu.

Meski X1 memiliki kontribusi variasi terbesar, nilai F-nya tidak melewati ambang signifikansi, sehingga model regresi tidak menyertakan variabel apapun pada tahap ini.

In [ ]:
# Menghitung nilai F hitung dan F tabel dari peubah dengan R tertinggi ke 2
print("")
print("TAHAP 3 - hitung nilai F dari X2")
df = n - (len(variable_terpilih)+ 1 + 1)
S2 = (JKT - JKR_sederhana[1]) / df
F_hitung = JKR_sederhana[1] / S2
F_tabel = st.f.ppf(0.95,1,df)

print("R =",f"{JKR_sederhana[1]:.4f}")
print("S kuadrat =",f"{S2:.4f}")
print("nilai F hitung =",f"{F_hitung:.4f}")
print("nilai F tabel =", f"{F_tabel:.4f}")

if F_hitung > F_tabel:
  print(f"peubah X{1+1} masuk dalam model")
  variable_terpilih.append(1)
else:
  print(f"peubah X{1+1} tidak masuk dalam model")


TAHAP 3 - hitung nilai F dari X2
R = 34089083106.4673
S kuadrat = 18101269101.9486
nilai F hitung = 1.8832
nilai F tabel = 4.0427
peubah X2 tidak masuk dalam model


Karena F hitung (1.8832) < F tabel (4.0427):

1. Variabel X2 (English) tidak signifikan secara statistik terhadap Salary.

2. X2 tidak dimasukkan ke dalam model regresi akhir.

Artinya, kemampuan Bahasa Inggris peserta tidak terbukti memengaruhi gaji secara langsung dan signifikan dalam model ini.

In [ ]:
# Menghitung nilai F hitung dan F tabel dari peubah dengan R tertinggi ke 3
print("")
print("TAHAP 3 - hitung nilai F dari X4")
df = n - (len(variable_terpilih)+ 1 + 1)
S2 = (JKT - JKR_sederhana[3]) / df
F_hitung = JKR_sederhana[3] / S2
F_tabel = st.f.ppf(0.95,1,df)

print("R =",f"{JKR_sederhana[3]:.4f}")
print("S kuadrat =",f"{S2:.4f}")
print("nilai F hitung =",f"{F_hitung:.4f}")
print("nilai F tabel =", f"{F_tabel:.4f}")

if F_hitung > F_tabel:
  print(f"peubah X{3+1} masuk dalam model")
  variable_terpilih.append(3)
else:
  print(f"peubah X{3+1} tidak masuk dalam model")


TAHAP 3 - hitung nilai F dari X4
R = 11103308928.5077
S kuadrat = 18580139397.3228
nilai F hitung = 0.5976
nilai F tabel = 4.0427
peubah X4 tidak masuk dalam model


Karena F hitung (0.5976) < F tabel (4.0427), maka:

1. Variabel X4 (ComputerProgramming) tidak signifikan secara statistik dalam memengaruhi variabel dependen (Salary).

2. X4 tidak dimasukkan ke dalam model regresi.

Artinya, kemampuan komputer pemrograman peserta tidak cukup kuat untuk menjelaskan perbedaan gaji secara statistik dalam data ini.

In [ ]:
# Menghitung nilai F hitung dan F tabel dari peubah dengan R tertinggi ke 4
print("")
print("TAHAP 3 - hitung nilai F dari X3")
df = n - (len(variable_terpilih)+ 1 + 1)
S2 = (JKT - JKR_sederhana[2]) / df
F_hitung = JKR_sederhana[2] / S2
F_tabel = st.f.ppf(0.95,1,df)

print("R =",f"{JKR_sederhana[2]:.4f}")
print("S kuadrat =",f"{S2:.4f}")
print("nilai F hitung =",f"{F_hitung:.4f}")
print("nilai F tabel =", f"{F_tabel:.4f}")

if F_hitung > F_tabel:
  print(f"peubah X{2+1} masuk dalam model")
  variable_terpilih.append(2)
else:
  print(f"peubah X{2+1} tidak masuk dalam model")


TAHAP 3 - hitung nilai F dari X3
R = 5854645065.4373
S kuadrat = 18689486561.1367
nilai F hitung = 0.3133
nilai F tabel = 4.0427
peubah X3 tidak masuk dalam model


Karena F hitung (0.3133) < F tabel (4.0427):

1. Variabel X3 (Logical Reasoning) tidak signifikan secara statistik untuk mempengaruhi Salary.

2. X3 tidak dimasukkan ke dalam model regresi.

Artinya, kemampuan logika peserta tidak berpengaruh secara bermakna terhadap gaji menurut data ini.

In [ ]:
# Menaikkan taraf signifikansi agar ada salah satu variabel yang masuk (dari 5% menjadi 10%)
print("")
print("TAHAP 3 - hitung nilai F")
df = n - (len(variable_terpilih)+ 1 + 1)
S2 = (JKT - JKR_sekarang) / df
F_hitung = JKR_sekarang / S2
F_tabel = st.f.ppf(0.90,1,df) # Menjadi 0.90

print("R =",f"{JKR_sederhana[idx_terbesar]:.4f}")
print("S kuadrat =",f"{S2:.4f}")
print("nilai F hitung =",f"{F_hitung:.4f}")
print("nilai F tabel =", f"{F_tabel:.4f}")

if F_hitung > F_tabel:
  print(f"peubah X{idx_terbesar+1} masuk dalam model")
  variable_terpilih.append(idx_terbesar)
else:
  print(f"peubah X{idx_terbesar+1} tidak masuk dalam model")


TAHAP 3 - hitung nilai F
R = 53710775862.0677
S kuadrat = 17692483836.2069
nilai F hitung = 3.0358
nilai F tabel = 2.8131
peubah X1 masuk dalam model


Peubah X1 (Graduation Year) masuk ke dalam model regresi karena nilai F hitung (3,0358) > F tabel (2,8131) pada taraf signifikansi 10%.

Ini berarti tahun kelulusan memiliki pengaruh terhadap gaji, meskipun tidak terlalu kuat.

Model regresi dapat dilanjutkan dengan X1 sebagai satu-satunya peubah yang signifikan.

**III. Uji Selanjutnya**

In [ ]:
idx_terbesar = np.argmax(JKR_sederhana)
selected_vars = [idx_terbesar]
remaining_vars = list(set(range(4)) - set(selected_vars))
JKR_terkini = JKR_sederhana[idx_terbesar]

df_error = n - (len(selected_vars) + 1)
S2 = (JKT - JKR_terkini) / df_error
Fh = JKR_terkini / S2
Ft = st.f.ppf(0.90, 1, df_error)
print(f"X{idx_terbesar+1} masuk model pertama: F hitung = {Fh:.4f} > F tabel = {Ft:.4f} \n")

# Tahap-tahap berikutnya
while len(selected_vars) < 4:
    best_1 = 0
    best_F = 0
    best_var = None
    best_JKR = None
    best_S2 = None
    for i in remaining_vars:
        vars_test = selected_vars + [i]
        JKR_new, _ = Jumlah_Kuadrat_Regresi_Berganda([kolom[j] for j in vars_test], kolom[-1])
        delta_JKR = JKR_new - JKR_terkini
        df_num = 1
        df_den = n - (len(vars_test) + 1)
        S2_new = (JKT - JKR_new) / df_den
        Fh = delta_JKR / S2_new
        Ft = st.f.ppf(0.90, df_num, df_den)
        print(f"Uji tambah X{i+1}: R = {delta_JKR:.4f}, F hitung = {Fh:.4f}, F tabel = {Ft:.4f}")
        if best_1 < delta_JKR:
          best_1 = delta_JKR
          if Fh > Ft and Fh > best_F:
              best_F = Fh
              best_var = i
              best_JKR = JKR_new
              best_S2 = S2_new

    if best_var is not None:
        selected_vars.append(best_var)
        remaining_vars.remove(best_var)
        JKR_terkini = best_JKR
        S2 = best_S2
        print(f"\n Peubah X{best_var+1} masuk model, F = {best_F:.4f}\n")
    else:
        print("\nTidak ada peubah signifikan lagi. Proses berhenti.\n")
        break

X1 masuk model pertama: F hitung = 3.0358 > F tabel = 2.8131 

Uji tambah X2: R = 27692187327.2812, F hitung = 1.5842, F tabel = 2.8154
Uji tambah X3: R = 17811369667.0532, F hitung = 1.0069, F tabel = 2.8154
Uji tambah X4: R = 6940759887.0275, F hitung = 0.3873, F tabel = 2.8154

Tidak ada peubah signifikan lagi. Proses berhenti.



Model regresi terbaik yang terbentuk hanya menggunakan satu peubah, yaitu X1 (Graduation Year).

Peubah X2 (English), X3 (Logical), dan X4 (Computer Programming) tidak memberikan kontribusi signifikan tambahan terhadap prediksi gaji (Salary).

In [ ]:
# Cetak model akhir
print("=== Model Akhir ===")
print("Peubah yang masuk model:", [f"X{i+1}" for i in selected_vars])

=== Model Akhir ===
Peubah yang masuk model: ['X1']


X1 (Graduation Year) adalah satu-satunya peubah yang signifikan dalam model regresi pada taraf signifikansi 10%.

Hubungan negatif antara X1 dan Salary mengindikasikan bahwa semakin baru tahun kelulusan, cenderung gaji awalnya sedikit lebih rendah.

**4. Penafsiran koefisien regresi**

In [ ]:
#menentukan koefisien regresi
JKR_final, beta = Jumlah_Kuadrat_Regresi_Berganda([kolom[i] for i in selected_vars], kolom[-1])
print(f"JKR Final : {JKR_final}")
print(f"Koefisien regresi: {beta}")
print(f"Persamaan regresi: Y = {beta[0]:.4f} {beta[1]:.4f} X1")

JKR Final : 53710775868.14893
Koefisien regresi: [ 6.15415862e+07 -3.04310345e+04]
Persamaan regresi: Y = 61541586.2102 -30431.0345 X1


Analisis Koefisien Regresi:
1. 61541586.2102 adalah konstanta/intersep (nilai Y saat X1 = 0).

2. -30431.0345 adalah koefisien regresi X1 (GraduationYear). Ini berarti jika tahun kelulusan naik 1 tahun, maka prediksi gaji (Salary) menurun sebesar 30.431 ribu atau 30,431 rupiah, asumsi variabel lain tetap.

Analisis Persamaan Regresi:
1. Setiap kenaikan 1 tahun pada variabel Tahun Kelulusan, akan menyebabkan rata-rata penurunan gaji sebesar Rp30.431,03, menurut model.

**5. Penafsiran nilai R-squared**

In [ ]:
#Menentukan R_Squared
R2 = JKR_final / JKT
print(f"Nilai R² = {R2:.4f}")

Nilai R² = 0.0595


Nilai R² = 0.0595 menunjukkan bahwa model kurang baik dalam menjelaskan gaji berdasarkan tahun kelulusan.

Artinya: hanya sekitar 5.95% variasi dalam gaji dapat dijelaskan oleh model regresi ini, yaitu yang hanya menggunakan Tahun Kelulusan (X1) sebagai prediktor.

Model regresi ini tidak cocok digunakan untuk prediksi atau pengambilan keputusan yang serius. Diperlukan peubah tambahan atau variabel baru yang lebih relevan agar model menjadi lebih akurat dan representatif.

**6. kesimpulan dan analisis tentang hubungan antara variabel, kelayakan model dikaitkan data**

1. Dari uji regresi tahap 1, semua variabel awalnya tidak signifikan secara statistik karena nilai F hitung lebih kecil dari F tabel, kecuali Graduation Year (X1) yang kemudian menjadi satu-satunya variabel yang masuk model saat F tabel diturunkan menjadi 2.8131.

2. Koefisien X1 (−30,430): Untuk setiap kenaikan satu tahun kelulusan (misalnya dari 2012 ke 2013), gaji cenderung berkurang sebesar 30,430 rupiah.
Artinya, lulusan baru cenderung digaji lebih rendah dibanding lulusan lama dalam dataset ini.

3. Nilai koefisien determinasi R² = 0.0595, yang artinya hanya sekitar 5.95% variasi gaji yang dapat dijelaskan oleh model ini, sementara sisanya sebesar 94.05% dipengaruhi oleh faktor lain yang tidak dimasukkan dalam model

4. Secara keseluruhan, tidak ada variabel prediktor dalam dataset ini yang memiliki hubungan kuat atau signifikan secara statistik dengan gaji. Hubungan antar variabel lebih kuat daripada hubungannya dengan gaji.

5. Model regresi yang dihasilkan sangat lemah dan tidak cukup baik untuk digunakan dalam memprediksi gaji berdasarkan variabel-variabel yang tersedia.